leopard.go

```diff
@@ -0,0 +1,846 @@
+package reedsolomon
+
+// This is a O(n*log n) implementation of Reed-Solomon
+// codes, ported from the C++ library https://github.com/catid/leopard.
+//
+// The implementation is based on the paper
+//
+// S.-J. Lin, T. Y. Al-Naffouri, Y. S. Han, and W.-H. Chung,
+// "Novel Polynomial Basis with Fast Fourier Transform
+// and Its Application to Reed-Solomon Erasure Codes"
+// IEEE Trans. on Information Theory, pp. 6284-6299, November, 2016.
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"math/bits"
+	"sync"
+	"unsafe"
+)
+
+// reedSolomonFF16 is like reedSolomon but for more than 256 total shards.
+type reedSolomonFF16 struct {
+	DataShards   int // Number of data shards, should not be modified.
+	ParityShards int // Number of parity shards, should not be modified.
+	Shards       int // Total number of shards. Calculated, and should not be modified.
+}
+
+// newFF16 is like New, but for more than 256 total shards.
+func newFF16(dataShards, parityShards int, opts ...Option) (*reedSolomonFF16, error) {
+	initConstants()
+
+	if dataShards <= 0 || parityShards < 0 {
+		return nil, ErrInvShardNum
+	}
+
+	if dataShards+parityShards > 65536 {
+		return nil, ErrMaxShardNum
+	}
+
+	r := &reedSolomonFF16{
+		DataShards:   dataShards,
+		ParityShards: parityShards,
+		Shards:       dataShards + parityShards,
+	}
+	return r, nil
+}
+
+type ffe uint16
+
+const (
+	bitwidth   = 16
+	order      = 1 << bitwidth
+	modulus    = order - 1
+	polynomial = 0x1002D
+)
+
+var (
+	fftSkew  *[modulus]ffe
+	logWalsh *[order]ffe
+)
+
+// Logarithm Tables
+var (
+	logLUT *[order]ffe
+	expLUT *[order]ffe
+)
+
+// Stores the partial products of x * y at offset x + y * 65536
+// Repeated accesses from the same y value are faster
+var mul16LUTs *[order]mul16LUT
+
+type mul16LUT struct {
+	LUT [4 * 16]ffe
+}
+
+func (r *reedSolomonFF16) Encode(shards [][]byte) error {
+	if len(shards) != r.Shards {
+		return ErrTooFewShards
+	}
+
+	if err := checkShards(shards, false); err != nil {
+		return err
+	}
+	return r.encode(shards)
+}
+
+func (r *reedSolomonFF16) encode(shards [][]byte) error {
+	shardSize := len(shards[0])
+	if shardSize%64 != 0 {
```

Comment:
minor nit, because a length of 0 % 64 == 0, so perhaps:
if shardSize == 0 || shardSize%64 != 0 {

Reply:
One of the few beauties of returning unexported types as interfaces is that people can't bypass the constructers for it. :)

Either way, I don't really see why we should restrict to multiple-of-64 lengths long term. If users want to be leopard C compatible, they will have to abide by the limit anyway, and we could add a non-default option to check for it.

The 64 byte multiple just seems like an over-eager limitation. It does simplify things, but ultimately it will just be annoying for users.